cmd/oras/root/manifest/index/create.go

```diff
@@ -0,0 +1,185 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package index
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras/cmd/oras/internal/argument"
+	"oras.land/oras/cmd/oras/internal/command"
+	"oras.land/oras/cmd/oras/internal/display"
+	"oras.land/oras/cmd/oras/internal/display/status"
+	oerrors "oras.land/oras/cmd/oras/internal/errors"
+	"oras.land/oras/cmd/oras/internal/option"
+	"oras.land/oras/cmd/oras/internal/output"
+	"oras.land/oras/internal/contentutil"
+	"oras.land/oras/internal/descriptor"
+	"oras.land/oras/internal/listener"
+)
+
+type createOptions struct {
+	option.Common
+	option.Target
+
+	sources   []string
+	extraRefs []string
+}
+
+func createCmd() *cobra.Command {
+	var opts createOptions
+	cmd := &cobra.Command{
+		Use:   "create [flags] <name>[:<tag[,<tag>][...]] [{<tag>|<digest>}...]",
+		Short: "[Experimental] Create and push an index from provided manifests",
+		Long: `[Experimental] Create and push an index from provided manifests. All manifests should be in the same repository
+
+Example - create an index from source manifests tagged 'linux-amd64' and 'linux-arm64', and push without tagging:
+  oras manifest index create localhost:5000/hello linux-amd64 linux-arm64
+
+Example - create an index from source manifests tagged 'linux-amd64' and 'linux-arm64', and push with the tag 'latest':
+  oras manifest index create localhost:5000/hello:latest linux-amd64 linux-arm64
+
+Example - create an index from source manifests using both tags and digests, and push with tag 'latest':
+  oras manifest index create localhost:5000/hello:latest linux-amd64 sha256:xxx
+
+Example - create an index and push it with multiple tags:
+  oras manifest index create localhost:5000/hello:tag1,tag2,tag3 linux-amd64 linux-arm64 sha256:xxx
+`,
+		Args: oerrors.CheckArgs(argument.AtLeast(1), "the destination index to create."),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			refs := strings.Split(args[0], ",")
+			opts.RawReference = refs[0]
+			opts.extraRefs = refs[1:]
+			opts.sources = args[1:]
+			if err := option.Parse(cmd, &opts); err != nil {
+				return err
+			}
+			// check that all tags given by the user are valid tags
+			allRefs := opts.extraRefs
+			if opts.Reference != "" {
+				allRefs = append(opts.extraRefs, opts.Reference)
+			}
+			for _, ref := range allRefs {
+				if !contentutil.IsValidTag(ref) {
+					return fmt.Errorf("%s is not a valid tag", ref)
+				}
+			}
+			return nil
```

Comment:
Let's do this check in another PR along with other commands.

Reply:
Please also open an issue to track this if there isn't one.

Reply:
See https://github.com/oras-project/oras/pull/1475#discussion_r1732233944 and #1479